src/internet_identity/src/anchor_management/registration/registration_flow_v2.rs

```diff
@@ -0,0 +1,172 @@
+use crate::anchor_management::registration::captcha::{
+    check_captcha_solution, create_captcha, make_rng, Base64,
+};
+use crate::anchor_management::registration::rate_limit::process_rate_limit;
+use crate::anchor_management::{activity_bookkeeping, post_operation_bookkeeping};
+use crate::state;
+use crate::state::flow_states::RegistrationFlowState;
+use crate::state::temp_keys::TempKeyId;
+use crate::storage::anchor::Device;
+use candid::Principal;
+use ic_cdk::api::time;
+use ic_cdk::caller;
+use internet_identity_interface::archive::types::{DeviceDataWithoutAlias, Operation};
+use internet_identity_interface::internet_identity::types::IdRegFinishError::IdentityLimitReached;
+use internet_identity_interface::internet_identity::types::{
+    CheckCaptchaArg, CheckCaptchaError, DeviceData, DeviceWithUsage, IdRegFinishArg,
+    IdRegFinishError, IdRegFinishResult, IdRegNextStepResult, IdRegStartError, IdentityNumber,
+    RegistrationFlowNextStep,
+};
+
+impl From<RegistrationFlowState> for RegistrationFlowNextStep {
+    fn from(value: RegistrationFlowState) -> Self {
+        match value {
+            RegistrationFlowState::CheckCaptcha {
+                captcha_png_base64, ..
+            } => RegistrationFlowNextStep::CheckCaptcha {
+                captcha_png_base64: captcha_png_base64.0,
+            },
+            RegistrationFlowState::FinishRegistration { .. } => RegistrationFlowNextStep::Finish,
+        }
+    }
+}
+
+pub async fn identity_registration_start() -> Result<IdRegNextStepResult, IdRegStartError> {
+    process_rate_limit().map_err(|_| IdRegStartError::RateLimitExceeded)?;
+
+    let caller = caller();
+    if caller == Principal::anonymous() {
+        return Err(IdRegStartError::InvalidCaller);
+    }
+
+    let (captcha_png_base64, flow_state) = captcha_flow_state(time()).await;
+    state::with_flow_states_mut(|flow_states| {
+        flow_states.new_registration_flow(caller, flow_state)
+    })
+    .map_err(|_| IdRegStartError::AlreadyInProgress)?;
+
+    Ok(IdRegNextStepResult {
+        next_step: RegistrationFlowNextStep::CheckCaptcha {
+            captcha_png_base64: captcha_png_base64.0,
+        },
+    })
+}
+
+async fn captcha_flow_state(flow_created_timestamp_ns: u64) -> (Base64, RegistrationFlowState) {
```

Comment:
Will you add that this returns `Option<Base64>` afterwards?

Reply:
You mean, once the captcha might be skipped? No, I think I will `identity_registration_start` with some logic to decide whether this function should be called or if the next step is `Finish`.